src/util/newconfig/ConfigDefinition.cpp

```diff
@@ -0,0 +1,213 @@
+     {"log_directory_max_size", ConfigValue{ConfigType::Integer}.defaultValue(50 * 1024)},
+     {"log_rotation_hour_interval", ConfigValue{ConfigType::Integer}.defaultValue(12)},
+     {"log_tag_style", ConfigValue{ConfigType::String}.defaultValue("uint")},
+     {"extractor_threads", ConfigValue{ConfigType::Integer}.defaultValue(2)},
+     {"read_only", ConfigValue{ConfigType::Boolean}.defaultValue(false)},
+     {"txn_threshold", ConfigValue{ConfigType::Integer}.defaultValue(0)},
+     {"start_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"finish_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_cert_file", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_key_file", ConfigValue{ConfigType::String}.optional()},
+     {"api_version.min", ConfigValue{ConfigType::Integer}},
+     {"api_version.max", ConfigValue{ConfigType::Integer}}}
+};
+
+ObjectView
+ClioConfigDefinition::getObject(std::string_view prefix, std::optional<std::size_t> idx) const
+{
+    std::string prefixWithDot = std::string(prefix) + ".";
+    for (auto const& [mapKey, mapVal] : map_) {
+        if (mapKey.starts_with(prefixWithDot))
+            ASSERT(!mapKey.ends_with(".[]"), "Trying to retrieve an object when value is an Array");
+
+        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<ConfigValue>(mapVal))
+            return ObjectView{prefix, *this};
+
+        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<Array>(mapVal)) {
+            ASSERT(std::get<Array>(mapVal).size() > idx, "index provided is out of scope");
+            return ObjectView{prefix, idx.value(), *this};
+        }
+    }
+    throw std::invalid_argument("Key is not found in config");
+}
+
+ArrayView
+ClioConfigDefinition::getArray(std::string_view prefix) const
+{
+    std::string key = std::string(prefix);
+    if (!prefix.contains(".[]"))
+        key += ".[]";
+
+    for (auto const& [mapKey, mapVal] : map_) {
+        if (mapKey.starts_with(key))
+            return ArrayView{key, *this};
+        ASSERT(!mapKey.starts_with(prefix), "Trying to retrieve an Array when value is an object");
+    }
+    throw std::invalid_argument("Key is not found in config");
+}
+
+ValueView
+ClioConfigDefinition::getValue(std::string_view fullKey) const
+{
+    if (map_.contains(fullKey) && std::holds_alternative<ConfigValue>(map_.at(fullKey))) {
+        return ValueView{std::get<ConfigValue>(map_.at(fullKey))};
+    }
+    ASSERT(
+        !map_.contains(fullKey) && std::holds_alternative<Array>(map_.at(fullKey)), "value of key is not Config Value."
+    );
+
+    throw std::invalid_argument("no matching key");
+}
+
+/**
+ * @brief Description of each config key and what they mean. Used to generate markdown file
+ *
+ * Key-value pairs. Key is configKey, value is its matching description
+ * Maybe_unused will be removed when markdown file is generated
+ */
+[[maybe_unused]] static constexpr ClioConfigDescription const DESCRIPTIONS{
```

Comment:
Can **ClioConfigDefinition** also be constexpr? If so, you can do the static check for their length, ensure their size are same. It will be nice to have such obligations.

Reply:
yea! The next step is to make it as constexpr as possible.